spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/repository/query/SpannerStatementQueryTests.java

```diff
@@ -0,0 +1,268 @@
+							statement.toString());
+					return null;
+				});
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void unspecifiedParametersTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"findTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[] {
+				"BUY",
+				"abcd",
+				"abc123",
+		};
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void unsupportedParamTypeTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"findTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[] {
+				"BUY",
+				"abcd",
+				"abc123",
+				8.88,
+				3.33,
+				new Trade(),
+				"ignored",
+		};
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = UnsupportedOperationException.class)
+	public void deleteTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn(
+				"deleteTop3DistinctIdActionPriceByActionAndSymbolOrTraderIdAndPriceLessThanOrPriceGreater"
+						+ "ThanEqualAndIdIsNotNullAndTraderIdIsNullOrderByIdDesc");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[0];
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test(expected = UnsupportedOperationException.class)
+	public void unSupportedPredicateTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn("countByTraderIdBetween");
+
+		SpannerOperations spannerOperations = mock(SpannerOperations.class);
+
+		PartTreeSpannerQuery partTreeSpannerQuery = new PartTreeSpannerQuery(Trade.class,
+				queryMethod, spannerOperations, new SpannerMappingContext());
+
+		Object[] params = new Object[0];
+
+		partTreeSpannerQuery.execute(params);
+	}
+
+	@Test
+	public void countTest() {
+
+		QueryMethod queryMethod = mock(QueryMethod.class);
+
+		when(queryMethod.getName()).thenReturn("countByAction");
```

Comment:
just for my sake, what does the "ByAction" part mean?

Reply:
the two words here are countBy and Action. countBy means to do a count, and Action means the Action property of the Trade. Because "operation" words (like IsNull or GreaterThan) are specified, it means the sole arg for the query method is the Action that you want to count.

Reply:
thanks for the explanation. I still don't fully understand, how would this translate to SQL? `SELECT COUNT(action) FROM trader` ? 
In that case, shouldn't we return 0 when a single `Trader()` is returned with a `null` action? 
[spanner docs for COUNT function with NULL values](https://cloud.google.com/spanner/docs/functions-and-operators#aggregate-functions)

Reply:
Ah, my misunderstanding: `By` is the separator for the `WHERE` clause. So this would translate to `SELECT COUNT(*) FROM trader where Action=<action_param>`.